common/src/main/java/com/wynntils/core/features/FeatureBase.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright © Wynntils 2022.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.features;
+
+import com.google.common.base.CaseFormat;
+import com.google.common.collect.ImmutableList;
+import com.wynntils.core.WynntilsMod;
+import com.wynntils.mc.utils.keybinds.KeyHolder;
+import com.wynntils.mc.utils.keybinds.KeyManager;
+import java.util.ArrayList;
+import java.util.List;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.TranslatableComponent;
+
+public abstract class FeatureBase extends Feature {
+    private boolean isListener = false;
+    private List<KeyHolder> keyMappings = new ArrayList<>();
+
+    /**
+     * Sets up this feature as an event listener. This should be called from the constructor.
+     * Behaviour if calling it after the feature is enabled is undefined.
+     */
+    public void setupEventListener() {
+        this.isListener = true;
+    }
+
+    /**
+     * Adds a keyHolder to the feature. This should be called from the constructor.
+     * Behaviour if calling it after the feature is enabled is undefined.
+     * @param keyHolder KeyHolder to add to the feature
+     */
+    public void setupKeyHolder(KeyHolder keyHolder) {
+        keyMappings.add(keyHolder);
+    }
+
+    protected String getNameCamelCase() {
+        String name = this.getClass().getTypeName().replace("Feature", "");
+        return CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_CAMEL, name);
+    }
+
+    @Override
+    public MutableComponent getNameComponent() {
+        return new TranslatableComponent("feature.wynntils." + getNameCamelCase() + ".name");
+    }
+
+    @Override
+    protected void onInit(ImmutableList.Builder<Condition> conditions) {}
```

Comment:
I may as well add a few comments on the `Condition` system I had added. Conditions were made to be like requirements needed to enable a feature, like an active api, needing champion/a rank, needing to be enabled by the user, etc. `onInit` could be renamed to show that it is only really used for adding conditions, and that could also be made a `setUp` method and abstracted away from feature.

About the `Condition` code itself, it feels clunky. `Condition` is an awkward inner class that needs to access methods of the surrounding `Feature`, 

```java
        if (this.conditions.isEmpty()) {
            enable();
        } else {
            this.conditions.forEach(Condition::init); //ends up calling `canEnable` if condition is immediately satisfied.
        }
```
, there should really be a better way to check whether all the conditions are satisfied than looping through them.

It'd be nice to deal with these issues at some time, but I think that Conditions work in some odd way.

Reply:
While reading this and seeing the refactors, do you think that being registered to the event listener could be a `Condition`? Same with having a registered keybind. I feel like if we cannot satisfy those conditions, most features would not function properly. So, perhaps they could be conditions?

Reply:
Also, we do have features that require the API but do not have a condition for them. 

Furthermore, let's say we have a feature that needs active API (with regular updates), like War Messages or Territory changes on map. Isn't it out of the scope of Conditions to see if the API is active and responsive? 

If not, should it prevent loading the feature, should it try to reload the API to see if it works or maybe inform the user on login that specific features were could not be loaded because conditions were not met?

Reply:
Can we leave conditions out of this PR? I don't really see how it fits in, and it can certainly be ripe for some redesign, but I have a limited time to spend on Wynntils, and I'm just worried this  PR gets stalled for too long, and that it will be tricky to keep it up with other changes.

Reply:
> While reading this and seeing the refactors, do you think that being registered to the event listener could be a `Condition`? Same with having a registered keybind. I feel like if we cannot satisfy those conditions, most features would not function properly. So, perhaps they could be conditions?

Yeah but if those don't function it requires a fix to the codebase, versus like an unresponsive api or missing permission.

> Also, we do have features that require the API but do not have a condition for them.

Which features?

> Furthermore, let's say we have a feature that needs active API (with regular updates), like War Messages or Territory changes on map. Isn't it out of the scope of Conditions to see if the API is active and responsive?

Conditions are imo suppose to be more of a loading feature thing, but that's a good concern. If it's an active API then the feature could deal with it itself.

> If not, should it prevent loading the feature, should it try to reload the API to see if it works or maybe inform the user on login that specific features were could not be loaded because conditions were not met?

It should notify the player that the api didn't work. /wynntils reload is suppose to be a reset mechanism to reinit everything, and should be suggested to try again.

Reply:
Also yeah this is out of scope